EIPS/eip-7399.md

```diff
@@ -0,0 +1,222 @@
+
+ The `amount` and `fee` need to be transferred to a `repayment receiver` before the end of the `flash` call. The `fee` can be set to zero `asset`.
+
+The `callback function` can return any arbitrary data which will be received by the `initiator` as the return value of the `flash` call.
+
+### Lender Specification
+
+A `lender` MUST implement the [ERC-7399](./eip-7399.md) interface.
+
+```
+// SPDX-License-Identifier: CC0
+pragma solidity >=0.6.0 <0.9.0;
+
+import { IERC20 } from "./IERC20.sol";
+
+interface IERC7399 {
+    /// @dev The amount of currency available to be lent.
+    /// @param asset The loan currency.
+    /// @return The amount of `asset` that can be borrowed.
+    function maxFlashLoan(
+        address asset
+    ) external view returns (uint256);
+
+    /// @dev The fee to be charged for a given loan. Returns type(uint256).max if the loan is not possible.
+    /// @param asset The loan currency.
+    /// @param amount The amount of assets lent.
+    /// @return The amount of `asset` to be charged for the loan, on top of the returned principal.
+    function flashFee(
+        IERC20 asset,
+        uint256 amount
+    ) external view returns (uint256);
+
+    /// @dev Initiate a flash loan.
+    /// @param loanReceiver The address receiving the flash loan
+    /// @param asset The asset to be loaned
+    /// @param amount The amount to loaned
+    /// @param data The ABI encoded user data
+    /// @param callback The address and signature of the callback function
+    /// @return result ABI encoded result of the callback
+    function flash(
+        address loanReceiver,
+        ERC20 asset,
+        uint256 amount,
+        bytes calldata data,
+        /// @dev callback. This is a combination of the callback receiver address, and the signature of callback
+        /// function. It is encoded packed as 20 bytes + 4 bytes.
+        /// @dev the return of the callback function is not encoded in the parameter, but must be `returns (bytes
+        /// memory)` for compliance with the standard.
+        /// @param initiator The address that called this function
+        /// @param paymentReceiver The address that needs to receive the amount plus fee at the end of the callback
+        /// @param asset The asset to be loaned
+        /// @param amount The amount to loaned
+        /// @param fee The fee to be paid
+        /// @param data The ABI encoded data to be passed to the callback
+        /// @return result ABI encoded result of the callback
+        function(address, address, IERC20, uint256, uint256, bytes memory) external returns (bytes memory) callback
+    )
+        external
+        returns (bytes memory);
+}
+
+```
+
+The `flashFee` function MUST return the fee charged for a loan of `amount` `asset`. If the asset is not supported `flashFee` MUST revert. If the lender doesn't have enough liquidity to loan `amount` the fee returned must be `type(uint256).max`.
```

Comment:
For some contracts, the difference between a "not supported" asset and an asset for which the lender has zero liquidity can be unclear. For example UniV4 singleton contract: it technically supports all assets, but it can have zero liquidity for some of them. So should it reverts or return `type(uint).max` in this case ?

Reply:
In that case I would return `type(uint).max`, since the asset is technically supported but has no liquidity.

The rationale is that borrowers will usually know which assets they can get from each lender, but call `maxFlashLoan` anyway to find out if they can borrow a particular amount at a particular time. Reverting on no liquidity would be bad UX.

Returning `type(uint).max` on `flashFee` follows the same logic. Borrowers are expected to know what they are doing and where to get each asset from. They can bypass the `maxFlashLoan` check and use `flashFee` instead to get both data points (enough liquidity and fee to pay) in a single call. A revert on `flashFee` is unexpected and notifies the borrower that they are doing something wrong.

Reply:
That said, this is a great point, and probably needs to be added to the EIP. In the [wrappers we have just released](https://github.com/alcueca/erc7399-wrappers) I'm not following a consistent approach on `flashFee` for Aave (probably needs to revert more), Balancer (shouldn't revert) and Uniswap (shouldn't revert).

Reply:
Fixed the EIP, thanks for your feedback 👍